### PR TITLE
GSYE-111: Validate the required energy of recommendations

### PR DESCRIPTION
**Pull request by Hassan754**
When the recommendation is created based on orders requirements, we have to use the values inside these requirements for validation Instead of validating the original members of the orders.

### CODE REVIEW

File: src/gsy_e/models/market/two_sided.py [hunk 1]
```diff
@@ -337,48 +330,59 @@ def match_recommendations(
...
+        selected_energy = recommendation.selected_energy
+        clearing_rate = recommendation.trade_rate
+        market_offer = self.offers.get(recommendation.offer["id"])
+        market_bid = self.bids.get(recommendation.bid["id"])
+
+        if not (market_offer and market_bid):
+            # If not all offers bids exist in the market, skip the current recommendation
+            raise InvalidBidOfferPairException("Not all bids and offers exist in the market.")
+        bid_energy = recommendation.bid_energy
+        offer_energy = market_offer.energy
```
Comment (hannesdiedrich) [inline src/gsy_e/models/market/two_sided.py:371]: As both variables are only used once in this method, you can omit declaring the additional variables here, right?

File: src/gsy_e/models/market/two_sided.py [hunk 2]
```diff
@@ -337,48 +330,59 @@ def match_recommendations(
...
         if selected_energy > bid_energy:
             raise InvalidBidOfferPairException(
                 f"Energy traded {selected_energy} is higher than bids energy {bid_energy}.")
```
Comment (hannesdiedrich) [inline src/gsy_e/models/market/two_sided.py:374]: Why are we returning the `bid_energy` here and the `offer_energy` a few lines below in the bid_energy_rate but don't do it in the error messages for the rates?

File: tests/test_two_sided_market.py [hunk 3]
```diff
@@ -391,56 +431,16 @@ def test_matching_list_gets_updated_with_residual_offers(self):
...
-    for p in patches:
-        p.start()
-    yield
-    for p in patches:
-        p.stop()
-
-
 class TestTwoSidedMarketMatchRecommendations:
     """Class Responsible for testing two sided market's matching functionality."""
-    def test_match_recommendations_fake_offer_bid(self, market, two_sided_market_matching):
```
Comment (hannesdiedrich) [inline tests/test_two_sided_market.py:411]: Why was this removed?

**Review by hannesdiedrich — COMMENTED** [review]
Two minor questions, LGTM once answered.

File: tests/test_two_sided_market.py (hunk 3 above)
Comment (Hassan754) [reply tests/test_two_sided_market.py:411]: I have moved the validation of existence of orders from the `match_recommendations` to the `validate_bid_offer_match` method, thus this test was removed and another test was added, please have a look at the `test_validate_bid_offer_match_orders_do_not_exist`

Comment (hannesdiedrich) [reply tests/test_two_sided_market.py:411]: I see, thanks and sorry for overseeing!

**Review by hannesdiedrich — APPROVED** [review]

File: tests/test_two_sided_market.py [hunk 4]
```diff
@@ -280,13 +315,14 @@ def test_market_trade_bid_partial(
...
+    @staticmethod
     def test_market_accept_bid_emits_bid_split_on_partial_bid(
-            self, called, market=TwoSidedMarket(bc=MagicMock(), time_slot=pendulum.now())):
+            called, market=TwoSidedMarket(bc=MagicMock(), time_slot=pendulum.now())):
         market.add_listener(called)
         bid = market.bid(20, 20, "A", "A")
         trade_offer_info = TradeBidOfferInfo(1, 1, 1, 1, 1)
         trade = market.accept_bid(bid, energy=1, trade_offer_info=trade_offer_info)
-        assert all([ev != repr(MarketEvent.BID_DELETED) for c in called.calls for ev in c[0]])
+        assert all((ev != repr(MarketEvent.BID_DELETED) for c in called.calls for ev in c[0]))
```
Comment (spyrostz) [inline tests/test_two_sided_market.py:325]: Minor, but I think that not even the second set of brackets is needed here.

**Review by spyrostz — APPROVED** [review]
One minor comment, LGTM

File: tests/test_two_sided_market.py (hunk 4 above)
Comment (Hassan754) [reply tests/test_two_sided_market.py:325]: Thanks! I'll fix this in my other in progress PR
